include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1466,6 +1466,119 @@ struct __partial_merge_kernel
     }
 };
 
+//Searching for an intersection of virtual matrix (n1, n2) diagonal with the Merge Path to define sub-ranges
+//to serial merge. For example, a virtual matrix for [0,1,1,2,3] and [0,0,2,3] shown below:
+//     0   1  1  2   3
+//    ------------------
+//   |--->
+// 0 | 0 | 1  1  1   1
+//   |   |
+// 0 | 0 | 1  1  1   1
+//   |   ---------->
+// 2 | 0   0  0  0 | 1
+//   |             ---->
+// 3 | 0   0  0  0   0 |
+template<typename Rng1, typename Rng2, typename _Index, typename _Index1, typename _Index2>
+auto __find_start_point(Rng1& __rng1, Rng2& __rng2, _Index __i_elem, _Index1 __n_1, _Index2 __n_2)
+{
+    _Index1 __end1 = 0;
+    _Index2 __end2 = 0;
+    if (__i_elem < __n_1) //a condition to specify upper or down part of the merge matrix to be processed
+    {
+        auto __q = __i_elem; //digonal index
+        auto __n_diag = __q < __n_2 ? __q : __n_2; //digonal size
+
+        oneapi::dpl::counting_iterator<_Index> __diag_it(0);
+        auto __res = std::lower_bound(__diag_it, __diag_it + __n_diag, 1, [&__rng1, &__rng2, __q](auto __v1, auto __v2)
+            {
+                auto __val1 = __rng1[__q - __v1 - 1] < __rng2[__v1];
+                return __val1 < __v2;
+            });
+        __end1 = __q - *__res;
+        __end2 = *__res;
+    }
+    else
+    {
+        auto __q = __i_elem - __n_1; //diagonal index
+
+        auto __n_diag = __n_2 - __q;
+        __n_diag = __n_diag < __n_1 ? __n_diag : __n_1; //digonal size
+
+        counting_iterator<_Index> __diag_it(0);
+        auto __res = std::lower_bound(__diag_it, __diag_it + __n_diag, 1, [&__rng1, &__rng2, __n_1, __q](auto __v1, auto __v2)
+            {
+                auto __val1 = __rng1[__n_1 - __v1 - 1] < __rng2[__q + __v1];
+                return __val1 < __v2;
+            });
+
+        __end1 = __n_1 - *__res;
+        __end2 = __q + *__res;
+    }
+    return std::make_pair(__end1, __end2);
+}
+
+template<typename _Rng1, typename _Rng2, typename _Rng3, typename _Index1, typename _Index2, typename _Index3,
+         typename _Size, typename _Size1, typename _Size2, typename _Compare>
+void
+__serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, _Index1 __start1, _Index2 __start2,
+               _Index3 __start3, _Size __chunk, _Size1 __n1, _Size2 __n2, _Compare __comp)
+{
+    if (__start1 >= __n1)
+    {
+        //copying a residual of the the second seq
+        const auto __n = std::min<_Index2>(__n2 - __start2, __chunk);
+        _ONEDPL_PRAGMA_UNROLL
+        for (auto __i = 0; __i < __n; ++__i)
```

Comment:
`__i` needs to be explicitly typed `__Index2` to accommodate `__n`.    This appears in a few more places below.


`_Size` should be included in the numbered `_SizeX`, types.

Reply:
You've changed the type to `::std::uint8_t`, which I believe will work if `__serial_merge` is called in the context of this algorithm.  However, if this is the intention, then we should probably just remove the `_Size` template argument, and just explicitly type `__chunk` as `::std::uint8_t`.  This solves the other discussion as well.

If instead we want `__serial_merge` to be general and available for use in other algorithms, then `__i` needs to be one of `_Size`, `_Index2` or `_Size2` to make sure there is no overflow.  Ideally it could be typed as the smallest of those types, and we could write some compile time helpers to accomplish that.